src/stats.ts

```diff
@@ -222,13 +244,35 @@ export abstract class StatsCommon<T extends number | bigint> implements Node.Sta
 	 */
 	public hasAccess(mode: number): boolean {
 		if (credentials.euid === 0 || credentials.egid === 0) {
-			//Running as root
+			// Running as root
 			return true;
 		}
 
-		// Mask for
-		const adjusted = (credentials.uid == this.uid ? S_IRWXU : 0) | (credentials.gid == this.gid ? S_IRWXG : 0) | S_IRWXO;
-		return (mode & this.mode & adjusted) == mode;
+		let perm = 0;
+
+		// Owner permissions
+		if (credentials.uid === this.uid) {
+			if (this.mode & S_IRUSR) perm |= R_OK;
+			if (this.mode & S_IWUSR) perm |= W_OK;
+			if (this.mode & S_IXUSR) perm |= X_OK;
+		}
+
+		// Group permissions
+		if (credentials.gid === this.gid) {
+			if (this.mode & S_IRGRP) perm |= R_OK;
+			if (this.mode & S_IWGRP) perm |= W_OK;
+			if (this.mode & S_IXGRP) perm |= X_OK;
+		}
+
+		// Others permissions
+		if (credentials.uid !== this.uid && credentials.gid !== this.gid) {
```

Comment:
This should always be applied. For example, a mode of 007 should grant rwx, but if the user or group id matches, it will not work correctly.

Reply:
I went ahead and just fixed this in [9dac60d](https://github.com/zen-fs/core/pull/131/commits/9dac60d95f2fb933f71e1941bcc30718fd3a3de9)

Reply:
It would be nice to have an unit test covering this scenario